From 9ce1c482b9d770e939062f892f6faa0641d6498b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jos=C3=A9=20Dapena=20Paz?= <jdapena@igalia.com>
Date: Sat, 14 Jun 2014 00:57:58 +0200
Subject: [PATCH] LTTNG tracing support in Qt Quick

:Release Notes:
LTTNG tracing support in Qt Quick, with misc tracepoints.

:Detailed Notes:
Added LTTNG tracing for QtQuick.

Added tracepoints for:
* Scenegraph rendering
* Material shader processing
* Texture uploads

:Testing Performed:

:QA Notes:

:Issues Addressed:
[GF-9999] Summary

Signed-off-by: Martin Jansa <Martin.Jansa@gmail.com>

Change-Id: Iadb777c9d8d34ad9bccd5e6582313467ef3f4cd6
---
 src/quick/items/qquickwindow.cpp                   |  4 ++++
 src/quick/quick.pro                                |  3 +++
 .../qsgdefaultdistancefieldglyphcache.cpp          | 14 ++++++++++++++
 src/quick/scenegraph/qsgdefaultrendercontext.cpp   |  4 ++++
 src/quick/scenegraph/util/qsgatlastexture.cpp      |  5 +++++
 src/quick/scenegraph/util/qsgtexture.cpp           |  4 ++++
 src/quick/scenegraph/util/qsgtexturematerial.cpp   |  3 +++
 7 files changed, 37 insertions(+)

diff --git a/src/quick/items/qquickwindow.cpp b/src/quick/items/qquickwindow.cpp
index b9d37daf7..b05c3afbb 100644
--- a/src/quick/items/qquickwindow.cpp
+++ b/src/quick/items/qquickwindow.cpp
@@ -84,6 +84,8 @@
 #include <private/qdebug_p.h>
 #endif
 
+#include <qtquick_tracepoints_p.h>
+
 QT_BEGIN_NAMESPACE
 
 Q_LOGGING_CATEGORY(DBG_TOUCH, "qt.quick.touch")
@@ -441,6 +443,7 @@ void forceUpdate(QQuickItem *item)
 
 void QQuickWindowPrivate::syncSceneGraph()
 {
+    Q_TRACE(qquickwindowprivate_syncscenegraph_entry);
     QML_MEMORY_SCOPE_STRING("SceneGraph");
     Q_Q(QQuickWindow);
 
@@ -476,6 +479,7 @@ void QQuickWindowPrivate::syncSceneGraph()
 
 void QQuickWindowPrivate::renderSceneGraph(const QSize &size)
 {
+    Q_TRACE(qquickwindowprivate_renderscenegraph_entry);
     QML_MEMORY_SCOPE_STRING("SceneGraph");
     Q_Q(QQuickWindow);
     if (!renderer)
diff --git a/src/quick/quick.pro b/src/quick/quick.pro
index 0f5f5abca..6eaea6efa 100644
--- a/src/quick/quick.pro
+++ b/src/quick/quick.pro
@@ -47,4 +47,7 @@ SOURCES += qtquick2.cpp
 # To make #include "qquickcontext2d_jsclass.cpp" work
 INCLUDEPATH += $$PWD
 
+TRACEPOINT_PROVIDER = $$PWD/qtquick.tracepoints
+CONFIG += qt_tracepoints
+
 load(qt_module)
diff --git a/src/quick/scenegraph/qsgdefaultdistancefieldglyphcache.cpp b/src/quick/scenegraph/qsgdefaultdistancefieldglyphcache.cpp
index c51358df7..b608fbb94 100644
--- a/src/quick/scenegraph/qsgdefaultdistancefieldglyphcache.cpp
+++ b/src/quick/scenegraph/qsgdefaultdistancefieldglyphcache.cpp
@@ -46,6 +46,7 @@
 #include <QtGui/private/qdistancefield_p.h>
 #include <QtGui/private/qopenglcontext_p.h>
 #include <QtQml/private/qqmlglobal_p.h>
+#include <qtquick_tracepoints_p.h>
 #include <qopenglfunctions.h>
 #include <qopenglframebufferobject.h>
 #include <qmath.h>
@@ -184,6 +185,7 @@ void QSGDefaultDistanceFieldGlyphCache::requestGlyphs(const QSet<glyph_t> &glyph
 
 void QSGDefaultDistanceFieldGlyphCache::storeGlyphs(const QList<QDistanceField> &glyphs)
 {
+    Q_TRACE(qsgdefaultdistancefieldglyphcache_storeglyphs_entry);
     typedef QHash<TextureInfo *, QVector<glyph_t> > GlyphTextureHash;
     typedef GlyphTextureHash::const_iterator GlyphTextureHashConstIt;
 
@@ -228,16 +230,20 @@ void QSGDefaultDistanceFieldGlyphCache::storeGlyphs(const QList<QDistanceField>
 #endif
         if (useTextureUploadWorkaround()) {
             for (int i = 0; i < glyph.height(); ++i) {
+                Q_TRACE(gltexsubimage2d_before)
                 m_funcs->glTexSubImage2D(GL_TEXTURE_2D, 0,
                                          c.x - padding, c.y + i - padding, glyph.width(),1,
                                          format, GL_UNSIGNED_BYTE,
                                          glyph.scanLine(i));
+                Q_TRACE(gltexsubimage2d_after)
             }
         } else {
+            Q_TRACE(gltexsubimage2d_before)
             m_funcs->glTexSubImage2D(GL_TEXTURE_2D, 0,
                                      c.x - padding, c.y - padding, glyph.width(), glyph.height(),
                                      format, GL_UNSIGNED_BYTE,
                                      glyph.constBits());
+            Q_TRACE(gltexsubimage2d_after)
         }
     }
 
@@ -275,6 +281,7 @@ void QSGDefaultDistanceFieldGlyphCache::createTexture(TextureInfo *texInfo,
                                                       int height,
                                                       const void *pixels)
 {
+    Q_TRACE(qsgdefaultdistancefieldglypcache_createtexture_entry);
     if (useTextureResizeWorkaround() && texInfo->image.isNull()) {
         texInfo->image = QDistanceField(width, height);
         memcpy(texInfo->image.bits(), pixels, width * height);
@@ -299,7 +306,9 @@ void QSGDefaultDistanceFieldGlyphCache::createTexture(TextureInfo *texInfo,
     const GLenum format = GL_ALPHA;
 #endif
 
+    Q_TRACE(glteximage2d_before);
     m_funcs->glTexImage2D(GL_TEXTURE_2D, 0, internalFormat, width, height, 0, format, GL_UNSIGNED_BYTE, pixels);
+    Q_TRACE(glteximage2d_after);
 
     texInfo->size = QSize(width, height);
 
@@ -319,6 +328,7 @@ static void freeFramebufferFunc(QOpenGLFunctions *funcs, GLuint id)
 
 void QSGDefaultDistanceFieldGlyphCache::resizeTexture(TextureInfo *texInfo, int width, int height)
 {
+    Q_TRACE(qsgdefaultdistancefieldglypcache_resizetexture_entry);
     QOpenGLContext *ctx = QOpenGLContext::currentContext();
     Q_ASSERT(ctx);
 
@@ -394,16 +404,20 @@ void QSGDefaultDistanceFieldGlyphCache::resizeTexture(TextureInfo *texInfo, int
 
         if (useTextureUploadWorkaround()) {
             for (int i = 0; i < texInfo->image.height(); ++i) {
+                Q_TRACE(gltexsubimage2d_before);
                 m_funcs->glTexSubImage2D(GL_TEXTURE_2D, 0,
                                          0, i, oldWidth, 1,
                                          format, GL_UNSIGNED_BYTE,
                                          texInfo->image.scanLine(i));
+                Q_TRACE(gltexsubimage2d_after);
             }
         } else {
+            Q_TRACE(gltexsubimage2d_before);
             m_funcs->glTexSubImage2D(GL_TEXTURE_2D, 0,
                                      0, 0, oldWidth, oldHeight,
                                      format, GL_UNSIGNED_BYTE,
                                      texInfo->image.constBits());
+            Q_TRACE(gltexsubimage2d_after);
         }
 
         m_funcs->glPixelStorei(GL_UNPACK_ALIGNMENT, alignment); // restore to previous value
diff --git a/src/quick/scenegraph/qsgdefaultrendercontext.cpp b/src/quick/scenegraph/qsgdefaultrendercontext.cpp
index 73b79c630..d13de4893 100644
--- a/src/quick/scenegraph/qsgdefaultrendercontext.cpp
+++ b/src/quick/scenegraph/qsgdefaultrendercontext.cpp
@@ -48,6 +48,8 @@
 #include <QtQuick/private/qsgcompressedtexture_p.h>
 #include <QtQuick/private/qsgdefaultdistancefieldglyphcache_p.h>
 
+#include <qtquick_tracepoints_p.h>
+
 QT_BEGIN_NAMESPACE
 
 #define QSG_RENDERCONTEXT_PROPERTY "_q_sgrendercontext"
@@ -176,6 +178,8 @@ static QBasicMutex qsg_framerender_mutex;
 
 void QSGDefaultRenderContext::renderNextFrame(QSGRenderer *renderer, uint fboId)
 {
+    Q_TRACE(qsgdefaultrendercontext_rendernextframe_entry);
+
     if (m_serializedRender)
         qsg_framerender_mutex.lock();
 
diff --git a/src/quick/scenegraph/util/qsgatlastexture.cpp b/src/quick/scenegraph/util/qsgatlastexture.cpp
index 921ed0c1f..5207c7ba8 100644
--- a/src/quick/scenegraph/util/qsgatlastexture.cpp
+++ b/src/quick/scenegraph/util/qsgatlastexture.cpp
@@ -59,6 +59,8 @@
 
 #include <private/qquickprofiler_p.h>
 
+#include <qtquick_tracepoints_p.h>
+
 QT_BEGIN_NAMESPACE
 
 #ifndef GL_BGRA
@@ -353,6 +355,7 @@ Texture *Atlas::create(const QImage &image)
 
 void Atlas::upload(Texture *texture)
 {
+    Q_TRACE(atlas_upload_entry);
     const QImage &image = texture->image();
     const QRect &r = texture->atlasSubRect();
 
@@ -391,6 +394,7 @@ void Atlas::upload(Texture *texture)
 void Atlas::uploadBgra(Texture *texture)
 {
     QOpenGLFunctions *funcs = QOpenGLContext::currentContext()->functions();
+    Q_TRACE(atlas_uploadbgra_entry);
     const QRect &r = texture->atlasSubRect();
     QImage image = texture->image();
 
@@ -454,6 +458,7 @@ void Atlas::uploadBgra(Texture *texture)
 void Atlas::generateTexture()
 {
     QOpenGLFunctions *funcs = QOpenGLContext::currentContext()->functions();
+    Q_TRACE(atlas_generatetexture_entry);
     funcs->glTexImage2D(GL_TEXTURE_2D, 0, m_internalFormat, m_size.width(), m_size.height(), 0, m_externalFormat, GL_UNSIGNED_BYTE, nullptr);
 
 #if 0
diff --git a/src/quick/scenegraph/util/qsgtexture.cpp b/src/quick/scenegraph/util/qsgtexture.cpp
index d33bdffe7..b8ecd270c 100644
--- a/src/quick/scenegraph/util/qsgtexture.cpp
+++ b/src/quick/scenegraph/util/qsgtexture.cpp
@@ -43,6 +43,7 @@
 #include <qmath.h>
 #include <private/qquickprofiler_p.h>
 #include <private/qqmlglobal_p.h>
+#include <qtquick_tracepoints_p.h>
 #include <QtGui/qguiapplication.h>
 #include <QtGui/qpa/qplatformnativeinterface.h>
 #if QT_CONFIG(opengl)
@@ -706,6 +707,7 @@ void QSGPlainTexture::bind()
 #if QT_CONFIG(opengl)
     QOpenGLContext *context = QOpenGLContext::currentContext();
     QOpenGLFunctions *funcs = context->functions();
+    Q_TRACE(qsgplaintexture_bind_entry);
     if (!m_dirty_texture) {
         funcs->glBindTexture(GL_TEXTURE_2D, m_texture_id);
         if (mipmapFiltering() != QSGTexture::None && !m_mipmaps_generated) {
@@ -839,7 +841,9 @@ void QSGPlainTexture::bind()
     Q_QUICK_SG_PROFILE_RECORD(QQuickProfiler::SceneGraphTexturePrepare,
                               QQuickProfiler::SceneGraphTexturePrepareSwizzle);
 
+    Q_TRACE(gltextimage2d_before);
     funcs->glTexImage2D(GL_TEXTURE_2D, 0, internalFormat, m_texture_size.width(), m_texture_size.height(), 0, externalFormat, GL_UNSIGNED_BYTE, tmp.constBits());
+    Q_TRACE(gltextimage2d_after);
 
     qint64 uploadTime = 0;
     if (profileFrames)
diff --git a/src/quick/scenegraph/util/qsgtexturematerial.cpp b/src/quick/scenegraph/util/qsgtexturematerial.cpp
index 7b1d5abb2..ed34d00f5 100644
--- a/src/quick/scenegraph/util/qsgtexturematerial.cpp
+++ b/src/quick/scenegraph/util/qsgtexturematerial.cpp
@@ -44,6 +44,8 @@
 # include <QtGui/qopenglfunctions.h>
 #endif
 
+#include <qtquick_tracepoints_p.h>
+
 QT_BEGIN_NAMESPACE
 
 #if QT_CONFIG(opengl)
@@ -80,6 +82,7 @@ void QSGOpaqueTextureMaterialShader::initialize()
 void QSGOpaqueTextureMaterialShader::updateState(const RenderState &state, QSGMaterial *newEffect, QSGMaterial *oldEffect)
 {
     Q_ASSERT(oldEffect == nullptr || newEffect->type() == oldEffect->type());
+    Q_TRACE(qsgopaquetexturematerialshader_updatestate_entry);
     QSGOpaqueTextureMaterial *tx = static_cast<QSGOpaqueTextureMaterial *>(newEffect);
     QSGOpaqueTextureMaterial *oldTx = static_cast<QSGOpaqueTextureMaterial *>(oldEffect);
 
